acceptancetests/assess_lxd_space.py

```diff
@@ -0,0 +1,179 @@
+    )
+
+__metaclass__ = type
+log = logging.getLogger("assess_lxd_space")
+
+
+def assert_initial_spaces(client):
+    """Initial spaces status after bootstrap should show as:
+       no spaces to display"""
+    # Run juju spaces --format json
+    # Bug 1704105: https://bugs.launchpad.net/juju/+bug/1704105
+    # merge_stderr need to be set to True for the initial assert
+    # because spaces list is empty at this point.
+    raw_output = client.get_juju_output(
+        'spaces', '--format', 'json', include_e=False, merge_stderr=True)
+    expected_spaces = 'no spaces to display'
+    if expected_spaces not in raw_output:
+        raise JujuAssertionError(
+            'Incorrect initial spaces status. Found: {}\nExpected: {}'.format(
+            raw_output, expected_spaces))
+    else:
+        log.info('Initial spaces status is: {}.'.format(expected_spaces))
+
+
+def get_subnets(client):
+    # Run juju subnets --format json
+    raw_output = client.get_juju_output(
+        'subnets', '--format', 'json', include_e=False)
+    try:
+        subnets_output = json.loads(raw_output)
+    except ValueError as e:
+        log.error('Invalid output from juju subnets:\n{}'.format(raw_output))
+        raise e
+    subnets_dict = subnets_output['subnets']
+    subnets_cidr_list = subnets_output['subnets'].keys()
+    return (subnets_dict, subnets_cidr_list)
+
+
+def assert_initial_subnets(client):
+    """There should be at least 1 subnet available once bootstrap is completed.
+       By default 4 subnets should be shown under an AWS controller."""
+    subnets_dict = get_subnets(client)[0]
+    if not subnets_dict:
+        raise JujuAssertionError('No subnet can be Found.')
+    else:
+        log.info('{} subnet(s) have been found.'.format(str(len(subnets_dict))))
+
+
+def add_space_with_existing_subnet(client, space_name):
+    # Run juju add-space <space_name> <subnet_cidr>
+    subnets_cidr_list = get_subnets(client)[1]
+    subnet_cidr = subnets_cidr_list[0]
+    # Bug 1704105, merge_stderr=True is required.
+    client.get_juju_output(
+        'add-space', space_name, subnet_cidr, include_e=False, merge_stderr=True)
+    return (space_name, subnet_cidr)
+
+
+def assert_added_space(client):
+    """Validate if configurations in new added space are as expected.
+       Five checkpoints:
+       1. After new space is added, juju spaces --format json should give output
+          in json format, in comparison to no space exists.
+       2. Total number of space should be 1.
+       3. The name of new added space should be the same as specified.
+       4. Total number of subnet CIDR from this new added space should be 1.
+       5. The CIDR data from new added space should be the same as specified.
+    """
+    space_name, subnet_cidr = add_space_with_existing_subnet(
+        client, space_name='testspace')
+    # New added space should exist at this point,
+    # set merge_stderr to False to eliminate noise in output.
```

Comment:
Since we have this stdout and stderr issue, I've tried my best to document it once I met. In this test script, merge_stderr has been shown multiple times and there is a detailed description for its first show.
"merge_stderr defaults to False", well, you and I know this, anyone else? I declare it here to alert user, and if user wants more info, simply Ctrl+F with merge_stderr as a keyword in this code file. Actually I will add more explanation here to tell user what those noise are.